[rebranch] Get rebranch building with new LLVM stable branch

AnthonyLatsis opened the pull request:
Coupled with https://github.com/swiftlang/llvm-project/pull/10531 and https://github.com/swiftlang/llvm-project/pull/10532.

AnthonyLatsis committed:
IRGen, LLVMPasses: Switch from obsoleted APIs to `llvm::getPointerAuthStableSipHash`

Per d68f8b96b1ff39cfcfcad94d5b16d28adaf6b5d0 (llvm-project).

AnthonyLatsis committed:
IRGen, LLVMPasses: Switch from obsoleted `nocapture` to `captures(none)`

Per 29441e4f5fa5f5c7709f7cf180815ba97f611297 (llvm-project).

AnthonyLatsis committed:
LLVMPasses: Switch from obsoleted `llvm::IRHash` to `llvm::stable_hash`

Per 1941c5180b91d792200d5e868d45c96e99bda35e (llvm-project).

AnthonyLatsis committed:
Adjust includes of moved LLVM header

Per 2ae968a0d9fb61606b020e898d884c82dd0ed8b5 (llvm-project).

AnthonyLatsis committed:
Refactor `llvm::opt::GenericOptTable` subclasses after changes to LLVM Option library

Per dd647e3e608ed0b2bac7c588d5859b80ef4a5976 (llvm-project).

AnthonyLatsis committed:
Refactor uses of obsoleted `llvm::DenseMap::{FindAndConstruct,getOrInsertDefault}`

Per 390b82dd4c485ec64cf8a6c52fb73e391792262e (llvm-project).

AnthonyLatsis committed:
Basic: Rename symbol in bitfield macro to not collide with Clang

Per 131acb07d814fabcc969dcaa63f4f352cd529267 (llvm-project).

AnthonyLatsis committed:
AST: Refactor use of obsoleted `llvm::Intrinsic::lookupLLVMIntrinsicByName`

Per f0297ae552e1e5aacafc1ed43968041994dc8a6e (llvm-project).

AnthonyLatsis committed:
AST: Conservatively handle new `llvm::Intrinsic::IITDescriptor::IITDescriptorKind` enum cases

Per 5a1e16f6de26c21cdfae1de05bd075d57029a3e1 (llvm-project).

AnthonyLatsis committed:
AST, ClangImporter, IRGen: Conservatively handle new HLSL `clang::BuiltinType::Kind` enum case

Per 52956b0f705499ae4a268d3629b402ecdc2cbbab (llvm-project).

AnthonyLatsis committed:
AST, ClangImporter: Fix `AMDGPU_TYPE` clang macro definitions

Per fe61dbf1d3350d11ab975e45f805f3bfe885b8b7 (llvm-project).

AnthonyLatsis committed:
ClangImporter: Conservatively handle new `clang::TargetInfo::BuiltinVaListKind` enum case

Per 2d6e7c2b359d4cafca8eaca4c9ed308a3a8fa6c1 (llvm-project).

AnthonyLatsis committed:
ClangImporter: Conservatively adjust uses of some Clang APIs (parameter added)

Per 08bda1cc6b0d2f1d31a89a76b5c154a11086c420 (llvm-project).

AnthonyLatsis committed:
ClangImporter: Adjust use of `clang::tooling::dependencies::ModuleDeps::CASFileSystemRootID` (type changed)

Per c3f10e5f166bb10afc091f21f549f0b1047701d0 (llvm-project).

AnthonyLatsis committed:
ClangImporter, Index: Refactor uses of obsoleted `clang::FileManager::getFile`

Per 99b1a2ac078fe52300d270b3e77ddbababa8f951 (llvm-project).

AnthonyLatsis committed:
ClangImporter, IDETool: Conservatively adjust uses of `clang::CompilerInstance::createDiagnostics` (parameter added)

Per df9a14d7bbf1180e4f1474254c9d7ed6bcb4ce55 (llvm-project).

AnthonyLatsis committed:
Adjust includes of moved Clang headers

Per 0d150db214e2aa13a825b563c7238e1243d61db1 (llvm-project).

AnthonyLatsis committed:
ClangImporter, IRGen: Conservatively handle new `clang::HLSLAttributedResourceType`

Per e00e9a3f8294c9b96cb0328bf136fab72aeec749 (llvm-project).

AnthonyLatsis committed:
IRGen: Conservatively handle new HLSL `clang::ParameterABI` enum cases

Per 89fb8490a99e612f7a574e8678b21a90f689f5b4 (llvm-project).

AnthonyLatsis committed:
ClangImporter: Refactor call to obsoleted `clang::ASTWriter::getMacroID`

Per 5d62a79bb79fee20f92f26dc55fd78440b9945ca (llvm-project).

AnthonyLatsis committed:
IRGen: Adjust uses of `clang::FieldDecl::getBitWidthValue` (parameter removed)

Per cfe26358e3051755961fb1f3b46328dc2c326895 (llvm-project).

AnthonyLatsis committed:
IRGen: Conservatively adjust use of `llvm::DIBuilder::createObjectPointerType` (parameter added)

Per a5fb2bbb2ad8488482843e2298fbe6f6a1d45bbd (llvm-project).

AnthonyLatsis committed:
DriverTool, IRGen: Adjust uses of `llvm::Module::setTargetTriple` (parameter type changed)

Per 979c275097a642e9b96c6b0a12f013c831af3a6e (llvm-project).

AnthonyLatsis committed:
IRGen: Adjust uses of `llvm::BasicBlock::getFirstNonPHIOrDbg` (return type changed)

Per 81d18ad86419fc612c7071e888d11aa923eaeb8a (llvm-project).

AnthonyLatsis committed:
IRGen: Adjust uses of `llvm::DIBuilder::insertDbgValueIntrinsic` (parameter type changed)

Per 1083ec647f16314bcc9af8c4d6b11f50d288bca6 (llvm-project).

AnthonyLatsis committed:
IRGen: Conservatively adjust uses of `llvm::PassBuilder::registerOptimizerLastEPCallback` (lambda parameter added)

Per 390300d9f41df6ad71f0f4196ef4885d4bd5dc48 (llvm-project).

AnthonyLatsis committed:
IRGen: Adjust use of `llvm::PassBuilder::buildO0DefaultPipeline` (parameter type changed)

Per 5445edb5d6667333f829f03932c2fc6a607e8306 (llvm-project).

AnthonyLatsis committed:
Immediate: Adjust uses of `llvm::jitlink::Symbol::{get,set}Name` (type signature changed)

Per 2ccf7ed277df28651b94bbee9fccefdf22fb074f (llvm-project).

AnthonyLatsis commented:
Waiting for cherry-picks to land before running tests.

Xazax-hun commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -4102,7 +4104,7 @@ void ClangModuleUnit::lookupAvailabilityDomains(
     return;
 
-  auto *varDecl = dyn_cast_or_null<clang::VarDecl>(domainInfo.Decl);
+  clang::VarDecl *varDecl = nullptr;
```
Is this change intended?

AnthonyLatsis replied:
This is temporary. I will ditch the "work around missing commit" commits as soon as the cherry-picks land on the new stable branch.

compnerd commented on include/swift/AST/Expr.h:
```diff
@@ -154,12 +154,12 @@ class alignas(8) Expr : public ASTAllocated<Expr> {
   );
 
-  SWIFT_INLINE_BITFIELD_FULL(CollectionExpr, Expr, 64-NumExprBits,
+  SWIFT_INLINE_BITFIELD_FULL(CollectionExpr, Expr, 64-NumberOfExprBits,
```
Could we perhaps namespace this as `swift::Expr::NumExprBits` or some such?

AnthonyLatsis replied:
I don’t think we can extrapolate this approach to the macros, though. The subclass bitfield macros have code that should expand to the parent class’s `Num*Bits`.

compnerd replied:
I'm not sure I follow. If the member is a `public: static const constexpr size_t NumExprBits = ...` that should be accessible by name for the sub types.

AnthonyLatsis replied:
Right, but it won’t prevent the preprocessor from expanding `Num##U##Bits` to whatever clang defined, will it?

bnbarham commented on lib/Option/Options.cpp:
```diff
@@ -16,15 +16,16 @@
 
-using namespace swift::options;
-using namespace llvm::opt;
+#define OPTTABLE_STR_TABLE_CODE
```
I swear every rebranch has a options table refactoring 😅

bnbarham commented on lib/LLVMPasses/ARCEntryPointBuilder.h:
```diff
@@ -252,11 +252,16 @@ class ARCEntryPointBuilder {
+    AttributeList AttrList;
+    AttrList = AttrList.addFnAttribute(C, Attribute::NoUnwind);
+    AttrList = AttrList.addParamAttribute(
+        C, 0, Attribute::getWithCaptureInfo(C, llvm::CaptureInfo::none()));
```
Is the 0 vs 1 here correct? (I haven't looked at the API change)

AnthonyLatsis replied:
Yes, `addParamAttribute` forwards to `addAttributeAtIndex` with index + 1.

AnthonyLatsis replied:
The gist of the API change is that we can no longer express `captures(none)` with just an attr kind.

bnbarham commented on lib/Immediate/SwiftMaterializationUnit.cpp:
```diff
@@ -197,11 +197,13 @@ renameFunctionBodies(llvm::orc::MaterializationResponsibility &MR,
+      const StringRef Name = *Sym->getName();
+
+      if (ToRename.count(Name)) {
         // FIXME: Get rid of the temporary when Swift's llvm-project is
         // updated to LLVM 17.
```
Can we do this now? CC @lhames

bnbarham commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -2438,7 +2438,7 @@ bool swift::writeEmptyOutputFilesFor(
       Context.getClangModuleLoader());
     llvmModule->setTargetTriple(
-      clangImporter->getTargetInfo().getTargetOpts().Triple);
+        llvm::Triple(clangImporter->getTargetInfo().getTargetOpts().Triple));
```
What was `getTargetOpts().Triple` changed to 🤔?

compnerd replied:
It is a `std::string` now I believe.

AnthonyLatsis replied:
It’s `llvm:Module::setTargetTriple` that switched to accepting a triple.

AnthonyLatsis commented:
@ahatanaka @rjmccall Could you proofread the ClangImporter and IRGen changes respectively please?

AnthonyLatsis committed:
SILGen, SILOptimizer: Fix assertion failures in `llvm::APInt` ctor

Constructing an unsigned 1-bit `APInt` with a value of -1 now hits an
assertion. Previously, the value would just be masked to become 1.
Fix this by constructing 1-bit `BuiltinIntegerType` literals with a
value of 1 instead of -1.
